### PR TITLE
FrameFile: Java 17 compatibility.

### PR DESCRIPTION
**Pull request by gianm**
DataSketches Memory.map is not Java 17 compatible, and from discussions
with the team, is challenging to make compatible with 17 while also
retaining compatibility with 8 and 11. So, in this patch, we switch away
from Memory.map and instead use the builtin JDK mmap functionality. Since
it only supports maps up to Integer.MAX_VALUE, we also implement windowed
mmaps in FrameFile, such that we can still handle large files.

Other changes:

1) Add two new "map" functions to FileUtils, which we use in this patch.
2) Add a footer checksum to the FrameFile format. Individual frames
   already have checksums, but the footer was missing one. Not directly
   related, but I thought of it since I was modifying the code that reads
   footers.

Unit tests use an artificially lower `maxMmapSize`, since I didn't think it was
going to be feasible to test with real 2GB+ files in unit tests.

This work is towards #12838.

### CODE REVIEW

**Comment by gianm** [comment]
The issues raised by LGTM are false alarms — the first two are about potential overflows that can't happen, and the last is about using task IDs in file paths, which is OK since they are path-safe. I pushed a commit to try to satisfy the alerts anyway.

File: processing/src/main/java/org/apache/druid/frame/file/FrameFile.java [hunk 1]
```diff
@@ -63,13 +57,18 @@
  *
  * - 2 bytes: {@link FrameFileWriter#MAGIC}
  * - NNN bytes: sequence of {@link FrameFileWriter#MARKER_FRAME} followed by one compressed frame (see {@link Frame})
+ * - 1 byte: {@link FrameFileWriter#MARKER_NO_MORE_FRAMES}
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/frame/file/FrameFile.java:60]: Should we also add a frameVersion byte?
We might add more things to this in the future then things like snapshotting for fault tolerance becomes tricky

**Review by cryptoe — APPROVED** [review]
Quite cool LGTM +1 non binding

File: processing/src/main/java/org/apache/druid/frame/file/FrameFile.java (hunk 1 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/frame/file/FrameFile.java:60]: We've got this covered: there's version bytes for each frame, & if we need to version the entire frame file format, we can increment the MAGIC.

File: processing/src/main/java/org/apache/druid/frame/file/FrameFile.java [hunk 2]
```diff
@@ -80,91 +79,173 @@ public class FrameFile implements Closeable
...
+  private final Memory footerMemory; // Footer is everything from the final MARKER_NO_MORE_FRAMES to EOF.
+  private final int maxMmapSize;
   private final int numFrames;
   private final int numPartitions;
   private final ReferenceCountingCloseableObject<Closeable> referenceCounter;
   private final Closeable referenceReleaser;
 
+  /**
+   * Mapped memory, starting from {@link #bufferOffset} in {@link #file}, up to max of {@link #maxMmapSize}. Acts as
+   * a window on the underlying file. Remapped using {@link #remapBuffer(long)}, freed using {@link #releaseBuffer()}.
```
[the remark this answers is not in the document]
Comment (gianm) [reply processing/src/main/java/org/apache/druid/frame/file/FrameFile.java:96]: My thinking is we don't need to block this patch on it, but yes, we should definitely add performance tests / benchmarks for the frame stuff generally.

**Review by vogievetsky — APPROVED** [review]
Yay for more Java support

**Pull request merged**